### PR TITLE
Add Physics-Informed Neural Operator (PINO) model and unit tests

### PR DESCRIPTION
**Pull request by DARHWOLF**
<h2>Description</h2>

<p>
  This PR introduces a new <strong>Physics-Informed Neural Operator (PINO)</strong> model to DeepChem,
  designed to solve partial differential equations (PDEs) such as the 1D Burgers’ equation by incorporating 
  physics-based residual loss and boundary condition enforcement. The implementation is based on the methodology 
  described in 
  <a href="https://arxiv.org/abs/2111.03794" target="_blank">
    Fourier Neural Operator for Parametric Partial Differential Equations
  </a>.
</p>

<p>
  The PINO model uses Fourier blocks to efficiently capture function representations and includes a mechanism 
  to learn physical parameters (via reparameterization using a learnable <code>alpha</code> so that 
  <code>nu = exp(alpha)</code> remains positive). In addition, comprehensive unit tests have been added to verify:
</p>

<ul>
  <li>Model initialization and architecture</li>
  <li>Multi-resolution input handling</li>
  <li>Boundary condition enforcement</li>
  <li>Physics compliance via loss reduction</li>
  <li>Parameter learning (ensuring the learned ν lies within an expected range)</li>
  <li>Autograd functionality for computing PDE residuals</li>
</ul>

<p>
The code has been tested locally in a Python 3.11 environment and all unit tests pass.
</p>

<h2>Type of change</h2>

<ul>
  <li>[ ] Bug fix (non-breaking change which fixes an issue)</li>
  <li>[x] New feature (non-breaking change which adds functionality)</li>
  <li>[ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)</li>
  <li>[ ] Documentation (modification for documents)</li>
</ul>

<h2>Checklist</h2>

<ul>
  <li>[x] My code follows <a href="https://deepchem.readthedocs.io/en/latest/development_guide/coding.html">the style guidelines of this project</a></li>
  <li>[x] Run <code>yapf -i deepchem/models/torch_models/pino.py</code> and <code>yapf -i deepchem/models/torch_models/tests/test_pino.py</code> (version 0.32.0)</li>
  <li>[x] Run <code>mypy -p deepchem</code> and check no errors (ignoring third-party stub issues)</li>
  <li>[x] Run <code>flake8 deepchem/models/torch_models/pino.py</code> and <code>flake8 deepchem/models/torch_models/tests/test_pino.py --count</code></li>
  <li>[ ] Run <code>python -m doctest deepchem/models/torch_models/pino.py</code> and check no errors</li>
  <li>[x] I have performed a self-review of my own code</li>
  <li>[x] I have commented my code, particularly in hard-to-understand areas</li>
  <li>[x] I have made corresponding changes to the documentation</li>
  <li>[x] I have added tests that prove my feature works</li>
  <li>[x] New unit tests pass locally with my changes</li>
  <li>[x] I have checked my code and corrected any misspellings</li>
</ul>

### CODE REVIEW

File: deepchem/models/torch_models/pino.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+from typing import Callable, Dict, Tuple
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+# --------------------- Model Components ---------------------
+class ActNorm(nn.Module):
+    def __init__(self, dim: int):
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/pino.py:31]: Please add documentation (including examples) and doctypes

File: deepchem/models/torch_models/pino.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+        super().__init__()
+        self.scale = nn.Parameter(torch.ones(1, 1, dim))
+        self.bias = nn.Parameter(torch.zeros(1, 1, dim))
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/pino.py:36]: Please add documentation

File: deepchem/models/torch_models/pino.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        out = x * self.scale + self.bias
+        return out
+
+
+class FourierBlock1D(nn.Module):
+    def __init__(self, in_channels: int, out_channels: int, modes: int):
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/pino.py:42]: Please add documentation (including examples) and doctypes

File: deepchem/models/torch_models/tests/test_pino.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        model = PINO(**self.common_args)
+        test_input = torch.tensor(
+            [[0.5, 0.1]], dtype=torch.float32, device=self.device, requires_grad=True
+        )
+        output = model.model(test_input)
+        loss = model._compute_pde_loss(test_input, output)
+        self.assertTrue(loss.requires_grad)
+        loss.backward()
+        self.assertIsNotNone(test_input.grad)
+        print("[TestPINO] test_autograd_physics_residual passed.")
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/tests/test_pino.py:168]: Please remove the print statements

File: deepchem/models/torch_models/tests/test_pino.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    (https://arxiv.org/abs/2111.03794)
+"""
+
+import pytest
+import unittest
+import numpy as np
+import torch
+import deepchem as dc  # noqa: F401 -- Not used directly here
+from deepchem.data import NumpyDataset
+from deepchem.models.torch_models.pino import PINO
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/tests/test_pino.py:18]: this has to be added to the try-except block, it will cause the Ci to break when torch is not installed

**Review by shreyasvinaya — CHANGES_REQUESTED** [review]
a lot of changes needed before it is ready to be merged in

File: deepchem/models/torch_models/pino.py [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+from deepchem.models.torch_models.torch_model import TorchModel
+from deepchem.data import Dataset
+from deepchem.trans import Transformer
+from typing import Any, Callable, Dict, Optional, Sequence, Tuple
+import logging
+
+# Configure logger for this module
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)  # Adjust as needed
+if not logger.handlers:
```
Comment (rbharath) [inline deepchem/models/torch_models/pino.py:25]: This is not standard; please match deepchem elsewhere

File: deepchem/models/torch_models/pino.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+    ch = logging.StreamHandler()
+    ch.setLevel(logging.DEBUG)
+    formatter = logging.Formatter(
+        "%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+    ch.setFormatter(formatter)
+    logger.addHandler(ch)
+
+
+# --------------------- Model Components ---------------------
+class ActNorm(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/pino.py:35]: All layers must have tests

Comment (rbharath) [reply deepchem/models/torch_models/pino.py:35]: Every layer must be documented in layers.rst
